File: c++/linx/gpio-libsimpleio.cpp
// LabView LINX GPIO using libsimpleio

// Copyright (C)2016-2017, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <libgpio.h>

#include "gpio-libsimpleio.h"

// GPIO pin constructor

GPIO_libsimpleio::GPIO_libsimpleio(int32_t pin)
{
  this->IsOutput = false;
  this->pin = pin;
  this->fd = -1;
}

// GPIO pin configuration method

void GPIO_libsimpleio::configure(int32_t direction, int32_t *error)
{
  GPIO_configure(this->pin, direction, 0, GPIO_EDGE_NONE, GPIO_ACTIVEHIGH, error);
  if (*error) return;

  this->IsOutput = direction;

  if (this->fd == -1)
  {
    int fd;

    GPIO_open(this->pin, &fd, error);
    if (*error) return;

    this->fd = fd;
  }
}

// GPIO pin read method

void GPIO_libsimpleio::read(int32_t *state, int32_t *error)
{
  if (this->fd == -1)
  {
    int fd;

    GPIO_configure(this->pin, GPIO_DIRECTION_INPUT, 0, GPIO_EDGE_NONE, GPIO_ACTIVEHIGH, error);
    if (*error) return;

    GPIO_open(this->pin, &fd, error);
    if (*error) return;

    this->IsOutput = false;
    this->fd = fd;
  }

  GPIO_read(this->fd, state, error);
}

// GPIO pin write method

void GPIO_libsimpleio::write(int32_t state, int32_t *error)
{
  if (this->fd == -1)
  {
    int fd;

    GPIO_configure(this->pin, GPIO_DIRECTION_OUTPUT, state, GPIO_EDGE_NONE, GPIO_ACTIVEHIGH, error);
    if (*error) return;

    GPIO_open(this->pin, &fd, error);
    if (*error) return;

    this->IsOutput = true;
    this->fd = fd;
  }

  GPIO_write(this->fd, state, error);
}

File: freepascal/objects/simpleio/gpio_libsimpleio.pas
{ GPIO pin services using libsimpleio                                         }

{ Copyright (C)2017-2023, Philip Munts dba Munts Technologies.                }
{                                                                             }
{ Redistribution and use in source and binary forms, with or without          }
{ modification, are permitted provided that the following conditions are met: }
{                                                                             }
{ * Redistributions of source code must retain the above copyright notice,    }
{   this list of conditions and the following disclaimer.                     }
{                                                                             }
{ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" }
{ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE   }
{ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  }
{ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE   }
{ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR         }
{ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF        }
{ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS    }
{ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN     }
{ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)     }
{ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE  }
{ POSSIBILITY OF SUCH DAMAGE.                                                 }

UNIT GPIO_libsimpleio;

INTERFACE

  USES
    libsimpleio,
    GPIO;

  TYPE
    Drivers = (PushPull, OpenDrain, OpenSource);

    Polarities = (ActiveLow, ActiveHigh);

    Edges = (None, Rising, Falling, Both);

    PinSubclass = CLASS(TInterfacedObject, GPIO.Pin)
      CONSTRUCTOR Create
       (desg     : libsimpleio.Designator;
        dir      : Direction;
        state    : Boolean = False;
        driver   : Drivers = PushPull;
        polarity : Polarities = ActiveHigh;
        edge     : Edges = None);

      CONSTRUCTOR Create
       (chip     : Cardinal;
        line     : Cardinal;
        dir      : Direction;
        state    : Boolean = False;
        driver   : Drivers = PushPull;
        polarity : Polarities = ActiveHigh;
        edge     : Edges = None);

      DESTRUCTOR Destroy; OVERRIDE;

      FUNCTION Read : Boolean;

      PROCEDURE Write(state : Boolean);

      PROPERTY state : Boolean READ Read WRITE Write;
    PRIVATE
      fd  : Integer;
      inp : Boolean;
      int : Boolean;
    END;

IMPLEMENTATION

  USES
    Errors,
    libGPIO;

  PROCEDURE ComputeFlags
   (dir        : Direction;
    driver     : Drivers;
    polarity   : Polarities;
    edge       : Edges;
    VAR flags  : Integer;
    VAR events : integer);

  BEGIN
    CASE dir OF
      GPIO.Input  : flags := LINE_REQUEST_INPUT;
      GPIO.Output : flags := LINE_REQUEST_OUTPUT;
    END;

    CASE driver OF
      PushPull    : flags := flags OR LINE_REQUEST_PUSH_PULL;
      OpenDrain   : flags := flags OR LINE_REQUEST_OPEN_DRAIN;
      OpenSource  : flags := flags OR LINE_REQUEST_OPEN_SOURCE;
    END;

    CASE polarity OF
      ActiveHigh  : flags := flags OR LINE_REQUEST_ACTIVE_HIGH;
      ActiveLow   : flags := flags OR LINE_REQUEST_ACTIVE_LOW;
    END;

    CASE edge OF
      None        : events := EVENT_REQUEST_NONE;
      Rising      : events := EVENT_REQUEST_RISING;
      Falling     : events := EVENT_REQUEST_FALLING;
      Both        : events := EVENT_REQUEST_BOTH;
    END;
  END;

  { GPIO_libsimpleio.PinSubclass constructors }

  CONSTRUCTOR PinSubclass.Create
   (desg     : libsimpleio.Designator;
    dir      : Direction;
    state    : Boolean;
    driver   : Drivers;
    polarity : Polarities;
    edge     : Edges);

  VAR
    flags  : Integer;
    events : Integer;
    error  : Integer;

  BEGIN
    ComputeFlags(dir, driver, polarity, edge, flags, events);

    libGPIO.LineOpen(desg.chip, desg.chan, flags, events, Ord(state),
      Self.fd, error);

    IF error <> 0 THEN
      RAISE GPIO.Error.Create('ERROR: libGPIO.LineOpen() failed, ' +
        StrError(error));

    Self.inp := (dir = GPIO.Input);
    Self.int := (Edge <> None);
  END;

  CONSTRUCTOR PinSubclass.Create
   (chip     : Cardinal;
    line     : Cardinal;
    dir      : Direction;
    state    : Boolean = False;
    driver   : Drivers = PushPull;
    polarity : Polarities = ActiveHigh;
    edge     : Edges = None);

  VAR
    flags  : Integer;
    events : Integer;
    error  : Integer;

  BEGIN
    ComputeFlags(dir, driver, polarity, edge, flags, events);

    libGPIO.LineOpen(chip, line, flags, events, Ord(state), Self.fd, error);

    IF error <> 0 THEN
      RAISE GPIO.Error.Create('ERROR: libGPIO.LineOpen() failed, ' +
        StrError(error));

    Self.int := (edge <> None);
  END;

  { GPIO_libsimpleio.PinSubclass destructor }

  DESTRUCTOR PinSubclass.Destroy;

  VAR
    error  : Integer;

  BEGIN
    libGPIO.Close(Self.fd, error);
    INHERITED;
  END;

  { GPIO read method }

  FUNCTION PinSubclass.Read : Boolean;

  VAR
    s      : Integer;
    error  : Integer;

  BEGIN
    IF Self.int THEN
      BEGIN
        libGPIO.LineEvent(Self.fd, s, error);

        IF error <> 0 THEN
          RAISE GPIO.Error.Create('ERROR: libGPIO.LineEvent() failed, ' +
            StrError(error))
      END
    ELSE
      BEGIN
        libGPIO.LineRead(Self.fd, s, error);

        IF error <> 0 THEN
          RAISE GPIO.Error.Create('ERROR: libGPIO.LineRead() failed, ' +
            StrError(error))
      END;

    Read := Boolean(s);
  END;

  { GPIO write method }

  PROCEDURE PinSubclass.Write(state : Boolean);

  VAR
    error  : Integer;

  BEGIN
    IF Self.inp THEN
      RAISE GPIO.Error.Create('ERROR: Cannot write to input pin');

    libGPIO.LineWrite(Self.fd, Ord(state), error);

    IF error <> 0 THEN
      RAISE GPIO.Error.Create('ERROR: libGPIO.LineWrite() failed, ' +
        StrError(error));
  END;

END.
